modules/protobuf/src/main/scala/smithy4s/protobuf/internals/TaggedCodec.scala

```diff
@@ -0,0 +1,555 @@
+/*
+ *  Copyright 2021-2024 Disney Streaming
+ *
+ *  Licensed under the Tomorrow Open Source Technology License, Version 1.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *     https://disneystreaming.github.io/TOST-1.0.txt
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package smithy4s.protobuf.internals
+
+import com.google.protobuf.CodedInputStream
+import com.google.protobuf.CodedOutputStream
+import smithy4s.Bijection
+import smithy4s.Lazy
+import smithy4s.protobuf.internals.TaggedCodec.FieldTags.OneOf
+import smithy4s.protobuf.internals.TaggedCodec.FieldTags.Simple
+import smithy4s.schema.CollectionTag
+
+import scala.collection.mutable.Buffer
+
+import TaggedCodec._
+
+private[protobuf] sealed trait TaggedCodec[A] {
+  def wireType: Int
+  def isPrimitive: Boolean
+  def prepareWrite(protoIndex: Int, a: A): WriteNode
+  def prepareNonEmptyWrite(protoIndex: Int, a: A): WriteNode
+  def prepareRead(): ReadNode[A]
+
+  final def imap[B](biject: Bijection[A, B]): TaggedCodec[B] =
+    TaggedCodec.IMapCodec(this, biject.to, biject.from)
+
+  final def imap[B](to: A => B, from: B => A): TaggedCodec[B] = this match {
+    case IMapCodec(underlying, to_, from_) =>
+      IMapCodec(underlying, to_.andThen(to), from.andThen(from_))
+    case other => IMapCodec(other, to, from)
+  }
+
+  // Adds a layer of message, where the field is indexed by 1.
+  final def wrap: MessageCodec[A] = {
+    val fieldTags = oneOfTags match {
+      case Some(tags) => FieldTags.OneOf(tags)
+      case None       => FieldTags.Simple(1)
+    }
+    MessageCodec(
+      IndexedSeq(FieldCodec(fieldTags, this, identity[A])),
+      seq => seq(0).asInstanceOf[A]
+    )
+  }
+
+  def oneOfTags: Option[Seq[Int]]
+
+}
+
+private[protobuf] object TaggedCodec {
+
+  final case class ScalarFieldCodec[A](
+      underlying: ScalarCodec[A]
+  ) extends TaggedCodec[A] {
+    def wireType: Int = underlying.wireType
+    def isPrimitive: Boolean = true
+    def oneOfTags: Option[Seq[Int]] = None
+    def write(protoIndex: Int, a: A, os: CodedOutputStream): Unit =
+      underlying.writeTag(protoIndex, a, os)
+    def prepareWrite(protoIndex: Int, a: A): WriteNode = {
+      if (a == underlying.zero) {
```

Comment:
**nitpick:** 
```suggestion
      if (underlying.isZero(a)) {
```

Reply:
there's another one in this file

Reply:
[c0cc265](https://github.com/disneystreaming/smithy4s/pull/1455/commits/c0cc265bb948ee389361829647eefdb363c8bd32)